rust/bridge/node/src/lib.rs

```diff
@@ -3,13 +3,289 @@
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+use async_trait::async_trait;
+use libsignal_bridge::node;
+use libsignal_protocol::*;
+use neon::context::Context;
 use neon::prelude::*;
+use signal_neon_futures::*;
+use std::fmt;
+use std::marker::PhantomData;
+use std::panic::AssertUnwindSafe;
+use std::sync::Arc;
 
 pub mod logging;
 
+#[derive(Debug)]
+struct CallbackError {
+    message: String,
+}
+
+impl CallbackError {
+    fn new(message: String) -> CallbackError {
+        Self { message }
+    }
+}
+
+impl fmt::Display for CallbackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "callback error {}", self.message)
+    }
+}
+
+impl std::error::Error for CallbackError {}
+
+fn js_error_to_rust(func: &'static str, err: String) -> SignalProtocolError {
+    SignalProtocolError::ApplicationCallbackError(func, Box::new(CallbackError::new(err)))
+}
+
+pub fn boxed_object<'a, T: 'static + Send>(
+    cx: &mut TaskContext<'a>,
+    value: T,
+) -> JsResult<'a, JsValue> {
+    Ok(cx.boxed(node::DefaultFinalize(value)).upcast())
+}
+
+struct NodeSenderKeyStore<'a> {
+    js_queue: EventQueue,
+    store_object: Arc<Root<JsObject>>,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> NodeSenderKeyStore<'a> {
+    fn new(cx: &mut FunctionContext, store: Handle<JsObject>) -> Self {
+        Self {
+            js_queue: cx.queue(),
+            store_object: Arc::new(store.root(cx)),
+            phantom: PhantomData,
+        }
+    }
+
+    async fn do_get_sender_key(
+        &self,
+        name: SenderKeyName,
+    ) -> Result<Option<SenderKeyRecord>, String> {
+        eprintln!("here...");
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
```

Comment:
If you use `into_inner` here, you don't have to `finalize` below (which is good panic protection).

Reply:
```
error[E0507]: cannot move out of an `Arc`
  --> rust/bridge/node/src/lib.rs:70:32
   |
70 |             let store_object = store_object_shared.into_inner(cx);
   |                                ^^^^^^^^^^^^^^^^^^^ move occurs because value has type `libsignal_bridge::node::Root<libsignal_bridge::node::JsObject>`, which does not implement the `Copy` trait
```

?

Reply:
Oops. >< My mistake and also something to think about how to simplify later.